llvm/lib/Target/AIE/AIEBaseAliasAnalysis.cpp

```diff
@@ -117,13 +152,281 @@ static const Value *getUnderlyingObjectAIE(const Value *V) {
+  IntrinsicChainInfo ChainInfo;
+  // We count backwards.
+  unsigned LastCalls = 0;
+  // Let's expose the real pointer.
+  TargetPointer = skipCast(TargetPointer);
+  for (unsigned Count = 0; Count < BaseObjectSearchLimit; ++Count) {
+
+    LastUpdate = skipCast(LastUpdate);
+    if (!LastUpdate->getType()->isPointerTy())
+      return std::nullopt; // not a pointer.
+
+    if (auto *GEP = dyn_cast<GEPOperator>(LastUpdate)) {
+      if (!GEP->hasAllZeroIndices())
+        return std::nullopt; // No pointer arithmetic is allowed.
+      // The result pointer and the first operand have
+      // the same value
+      LastUpdate = GEP->getPointerOperand();
+    } else if (auto *PHI = dyn_cast<PHINode>(LastUpdate)) {
+      // Reached final destination.
+      ChainInfo.IntrinsicCalls = ChainInfo.TotalIntrinsicCalls - LastCalls;
+      return ChainInfo;
+    } else if (auto *AIEObject = lookThroughAIEAddressingModes(LastUpdate)) {
+      IntrinsicArgsInfo NewArgSet = collectIntrinsicOperands(LastUpdate);
+      // Initialize counters.
+      if (!ChainInfo.IntrinsicParameters.CounterArgSet.empty() &&
+          !isAddressingIntrinsicOutput(
+              ArrayRef(ChainInfo.IntrinsicParameters.CounterArgSet),
+              AIEObject)) {
+        // if we already have counters, we need to reach the same
+        // AIEObject from the counters. Otherwise we can have some undesired
+        // counter update in between.
+        return std::nullopt; // not reachable.
+      }
+
+      if (!ChainInfo.IntrinsicParameters.ArgSet.empty() &&
+          ChainInfo.IntrinsicParameters.ArgSet != NewArgSet.ArgSet) {
+        // Constant iterator information mismatch between sequence of
+        // intrinsic calls. We stop here.
+        return std::nullopt;
+      }
+
+      // Update params for the next round.
+      ChainInfo.IntrinsicParameters.CounterArgSet = NewArgSet.CounterArgSet;
+      ChainInfo.IntrinsicParameters.ArgSet = NewArgSet.ArgSet;
+
+      LastUpdate = AIEObject;
+      ChainInfo.TotalIntrinsicCalls++;
+    } else {
+      return std::nullopt; // something that we can't understand.
+    }
+    // This information is crucial to calculate the number
+    // of pointer updates before TargetPointer.
+    if (LastUpdate == TargetPointer)
+      LastCalls = ChainInfo.TotalIntrinsicCalls;
+  }
+  return std::nullopt;
+}
+
+static AliasResult aliasAIEIntrinsic(const Value *ValueA, const Value *ValueB,
+                                     const Value *BaseA, const Value *BaseB) {
+
+  const PHINode *PhiA = dyn_cast<PHINode>(BaseA);
+  const PHINode *PhiB = dyn_cast<PHINode>(BaseB);
+
+  // Both bases must be PHI nodes.
+  if (!PhiA || !PhiB)
+    return AliasResult::MayAlias;
+
+  // Skip non-pointer values (cast from pointers?).
+  if (!PhiA->getType()->isPointerTy() || !PhiB->getType()->isPointerTy())
+    return AliasResult::MayAlias;
+
+  // Our target is loop-recurrent PHI nodes.
```

Comment:
I'm not sure whether the condition guarantees that they are loop-recurrent. I'm not sure whether you rely on it. In particular, GetExternalIndex and GetInternalIndex seem to assume that the phi node resides in a loop header.
It is not clear to me whether ValueA and ValueB are supposed to be in the same block as BaseA and BaseB.

Reply:
I think we should then also check that the incoming BBs are the same, and one of them is block itself where the PHI nodes are?